### ISSUE TITLE
Hexagon backend miscompiles non-native vector width interleaving and deinterleaving

### ISSUE THREAD

**Issue by dsharletg** [halide/Halide#1582]
When compiling a widening operation where the input is not a multiple of the native vector size, or a narrowing operation where the output is not a multiple of the native vector size, results in a miscompile.

For example, consider a narrowing cast that interleaves the operand, vectorized for one vector of input (half a vector of output):

```
y = narrow(x)
```

To make this operation correct, we deinterleave x:

```
y = narrow(deinterleave(x))
```

`deinterleave` only works on pairs of native vectors, as does narrow. The actual sequence of operations that occurs is:

```
y = narrow(pad(slice(deinterleave(pad(x))))
```

Where `pad` and `slice` either double the vector width by padding with undef, or slicing in half.

The problem is that the `pad(slice(.))` step throws away useful vector lanes, replacing them with undef, which can lead to miscompiles.

**Comment by ckun5** [comment]
I met this error when I scheduled with hexagon:

Error at
 ～/Qualcomm/Hexagon_SDK/3.3.0/tools/HEXAGON_Tools/8.1.04/Tools/Halide/tools/GenGen.cpp:4:
Interleave or deinterleave will result in miscompilation, see https://github.com/halide/Halide/issues/1582

I did as you said that to narrow the input as beblow:
       Img(w,h) = narrow(pad(slice(deinterleave(pad(input(w,h)))));
but it tells me:
       error: ‘pad’ was not declared in this scope
       error: ‘deinterleave’ was not declared in this scope
       error: ‘slice’ was not declared in this scope
       error: ‘pad’ was not declared in this scope
       error: redeclaration of ‘<typeprefixerror>pad’
       error: ‘narrow’ was not declared in this scope

So how can deal with that the input size  is not a multiple of 64/128?

**Comment by dsharletg** [comment]
Sorry, that was pseudo-code to explain the problem, not a fix.

The fix is to vectorize your code such that all expressions involved are one native vector (i.e. 64 or 128 bytes, for HVX64/HVX128).

If your data is not a multiple of this vector size, you can:

- Use TailStrategy::GuardWithIf when you vectorize. This will hopefully generate semi-vectorized code in the last vector (vectorized arithmetic, scalarized load/store).
- Use TailStrategy::ShiftInwards. This is makes it hard to generate code that assumes the data is aligned though.
- Pad the input and/or output so it is a multiple of the vector size. This is what we do in practice.

**Comment by dsharletg** [comment]
Also: if you have a large pipeline, you only need to deal with this for code that accesses input or output buffers. For the internal pipeline stages, use TailStrategy::RoundUp. Halide will pad these internal allocations as necessary.

**Comment by ckun5** [comment]
ok, thank you!

**Comment by shubhamp-ca** [comment]
This is an especially common issue even for casts of one native vector width.

Consider the following Func which is a 16-bit widening cast vectorized by a factor of 64.

`casted = <int16_t>(u8*2)`

The issue is seen here because, the above expr is pattern matched and replaced by an interleave of widening multiply of half-native vectors. There are no hexagon intrinisics for handling half-native width multiply ops, so we correctly assert whether the interleave op is a multiple of 2*native vector width.

The way around this for now, is to vectorize this by a factor 128 (double-vector width - i16x128).

It should be reasonable to vectorize by one-vector width, though. I am exploring a solution where if the interleave op is exactly 1 vector width (half-native operands), we check if the operand to be interleaved is one of the supported implicit widening intrinsics (widening multiplies, accumulating mpys etc.), and replace the intrinsic Call with corresponding Op (Mul in this case) while removing the interleave. In a nutshell, sort of "undo" the intrinsic pattern matched.

This won't resolve the non-native vector width interleave/deinterleaves but reduce the number of cases where this issue is seen. This curious issue is seen a lot in auto-scheduler, where the auto-scheduler tries to vectorize a stage by one-vector width and leads to compile-time error.

**Comment by dsharletg** [comment]
> I am exploring a solution where if the interleave op is exactly 1 vector width (half-native operands), we check if the operand to be interleaved is one of the supported implicit widening intrinsics (widening multiplies, accumulating mpys etc.), and replace the intrinsic Call with corresponding Op (Mul in this case) while removing the interleave. In a nutshell, sort of "undo" the intrinsic pattern matched.

This made me think: I think the issue is not necessarily about the interleaving/deinterleaving, but about the EliminateInterleaves operation. You could try removing the EliminateInterleaves pass on a case that currently fails, and see what happens.

If this fixes the issue, the solution might be to figure out a way to make EliminateInterleaves more conservative (don't push interleaves through operations unless they are of suitable vector width).

**Comment by shubhamp-ca** [comment]
Removing EliminateInterleaves isn't fixing the issue in this case and that is because the pattern is not introduced because of EliminateInterleaves pushing the interleaves, but rather, OptimizePatterns shouldn't have matched half-native vector multiplication to hexagon.mpy intrinsic.

OptimizePattern actually might be a good place to address this, i.e. before applying patterns, check vector width and only then apply Intrinsic patterns. Presently, we only check if it is a vector type before matching and applying patterns.

**Comment by dsharletg** [comment]
Fixed by #4165 !
